### ISSUE TITLE
🐛Bug and Formula Discrepancy in [utils.py/evaluation_sd]

### ISSUE THREAD

**Issue by bhc6** [kmc0207/Stableprompt#4]
### 1.Core Formula Discrepancy

The most significant issue appears to be a potential mismatch between the mathematical definition of `softmax_diff` as described in the paper and its actual implementation. If the core definition is incorrect, it fundamentally changes the model's objective function or the meaning of the reported results.

**Request:** Could the authors please clarify the exact mathematical formula/logic for the reward calculation (the `softmax_diff` value), and confirm the function's implementation is faithful to the paper's definition?

* **Reasoning:** Since `softmax_diff` is derived from `mean_reward = reward.mean().cpu()` in `evaluation_soft`, it represents the *mean* for that batch. To correctly calculate the *total mean* across all batches, we must accumulate the *total sum* of the reward, not just the mean itself. And this implementation also exists in many other evaluation functions of the same type.

* **Proposed Fix:**
    The accumulation line in evaluation_sd of utils.py should be changed to multiply by the batch size:
    ```python
    sd += softmax_diff[0] * batch_size 
    ```

Thank you for your valuable work and for reviewing these critical points to ensure the research is fully reproducible!

utils.py
```
def evaluation_sd(
    prompts,
    dataset,
    model,
    tokenizer,
    device,
    verbalizer=['Yes', 'No', 'Maybe'],
    dataset_size=100,
    debug=False,
    side='First',
    MaskLM=False,
    batch_size=4,
    soft_diff=False,
):
    dataloader = DataLoader(dataset,
                            batch_size=batch_size,
                            shuffle=False,
                            drop_last=False)
    model.eval()
    accuracys = []
    wrongs = []
    sds = []
    for prompt in prompts:
        total = 0
        correct = 0
        sd = 0
        with torch.no_grad():
            for batch in dataloader:
                if 'text' in batch.keys():
                    inputs = batch['text']
                else:
                    inputs = batch['sentence']
                targets = batch['label']
                softmax_diff, acc = evaluation_soft(
                    [prompt],
                    inputs,
                    targets,
                    model,
                    tokenizer,
                    device,
                    verbalizer,
                    side=side,
                )
                batch_size = len(targets)
                correct += acc[0] * batch_size
                total += batch_size
                sd += softmax_diff[0] # should * batch_size here just like “correct += acc[0] * batch_size”
            if debug:
                print(inputs, targets, acc)
        accuracy = correct / total
        soft_diff = sd / total # softmax_diff is calculated with mean_reward = reward.mean().cpu() it should be handled in the same way as accuracy
        accuracys.append(torch.Tensor([accuracy]))
        sds.append(torch.Tensor([soft_diff]))

    return accuracys, sds
```

```
def evaluation_soft(
    prompts,
    inputs,
    targets,
    model,
    tokenizer,
    device,
    verbalizer,
    debug=False,
    Fail_coefficient=1,
    Success_coefficient=1,
    return_reward=False,
    side='First',
):

    def _format_prompts(prompts, inputs, side):
        if side == 'First':
            template = "{prompt} Input : {sentence_1} Output:"
        else:
            template = "{sentence_1} {prompt}"
        return [
            template.format(sentence_1=s_1, prompt=prompt)
            for s_1, prompt in zip(inputs, prompts)
        ]

    def _get_next_token_index(input_ids):
        return input_ids.shape[1] - 1

    def _get_logits(texts, tokenizer, model, device):
        batch_size = len(texts)
        encoded_inputs = tokenizer(texts,
                                   padding='longest',
                                   truncation=True,
                                   return_tensors="pt",
                                   add_special_tokens=True)
        token_logits = model(**encoded_inputs.to(device)).logits
        next_token_indices = _get_next_token_index(encoded_inputs['input_ids'])
        out_logits = token_logits[range(batch_size), next_token_indices, :]
        return out_logits

    accuracies = []
    rewards = []
    model.eval()
    verbalizer_ids = tokenizer.convert_tokens_to_ids(verbalizer)
    batch_size = targets.size(0)
    
    for prompt in prompts:
        #Get logits
        current_prompts = [prompt for _ in range(batch_size)]
        formatted_templates = _format_prompts(current_prompts,
                                              inputs,
                                              side=side)
        all_logits = _get_logits(formatted_templates, tokenizer, model, device)

        #Get verbalizer logits
        verbalizer_logits = all_logits[:, verbalizer_ids]
        log_probs = F.softmax(verbalizer_logits, dim=1)
        #print(log_probs)
        #Get accuracy
        preds = torch.argmax(log_probs, dim=1).cpu()
        correct_predictions = torch.sum(preds == targets)
        accuracy = correct_predictions.item() / batch_size
        accuracies.append(accuracy) # This calculates the average accuracy of a prompt over a batch.

        reward = get_reward(all_logits,
                            targets,
                            Fail_coefficient=Fail_coefficient,
                            Success_coefficient=Success_coefficient)
        mean_reward = reward.mean().cpu() # This section calculates the average score (soft_diff) of a prompt across a batch.
        rewards.append(mean_reward)

    z_scaled_reward = rewards

    if return_reward:
        return z_scaled_reward, accuracies, rewards
    else:
        return z_scaled_reward, accuracies
```
